pymodm/base/models.py

```diff
@@ -388,7 +396,16 @@ def __repr__(self):
 
     def __eq__(self, other):
         if isinstance(other, MongoModelBase):
-            return self._data == other._data
+            if self._data == other._data:
+                return True
+            fields_self = set(iter(self))
+            fields_other = set(iter(other))
+            if fields_self != fields_other:
+                return False
+            for fname in fields_self:
+                if getattr(self, fname) != getattr(other, fname):
+                    return False
+            return True
```

Comment:
Create [PYMODM-78](https://jira.mongodb.org/browse/PYMODM-78)

Reply:
Unfortunately, it seems like we cannot simply keep the old behavior because of how the `LazyDecoder` pattern works. Consider:
```
    def test_embedded_document(self):
        contr = Contributor(name='Shep')
        # embedded field is not required.
        contr.full_clean()
        contr.save()

        # Attach an image.
        thumb = Image(image_url='/images/shep.png', alt_text="It's Shep.")
        contr.thumbnail = thumb
        contr.save()

        self.assertEqual(thumb, next(Contributor.objects.all()).thumbnail)
```

In the model instantiated from the database, all data are stored in the `_mongo_data` dictionary (since it is populated from the incoming SON from the DB). On the other hand, the information in `thumb` is stored in the `_python_data` dictionary. Therefore, this comparison ends up failing unless we explicitly access all the model attributes to ensure they are cast into Python which is obviously a non-intuitive pattern. 

Instead, what I'll do for now is improve the `LazyDecoder.__eq__` to at least combine the two dictionaries and perform a comparison. Its nowhere close to perfect, but should do for now.